# **2.Cargar datos:**

In [1]:
import pandas as pd


In [2]:
# utilizo el comando .read_csv para importar el archivo .csv.

amazon = pd.read_csv(r'amazon.csv')


# **3.Explorar y Preparar los datos:**

In [3]:
# analizo las 5 primeras filas
amazon.head()

product_id                                       product_name  \
0  B07JW9H4J1  Wayona Nylon Braided USB to Lightning Fast Cha...   
1  B098NS6PVG  Ambrane Unbreakable 60W / 3A Fast Charging 1.5...   
2  B096MSW6CT  Sounce Fast Phone Charging Cable & Data Sync U...   
3  B08HDJ86NZ  boAt Deuce USB 300 2 in 1 Type-C & Micro USB S...   
4  B08CF3B7N1  Portronics Konnect L 1.2M Fast Charging 3A 8 P...   

                                            category discounted_price  \
0  Computers&Accessories|Accessories&Peripherals|...             ₹399   
1  Computers&Accessories|Accessories&Peripherals|...             ₹199   
2  Computers&Accessories|Accessories&Peripherals|...             ₹199   
3  Computers&Accessories|Accessories&Peripherals|...             ₹329   
4  Computers&Accessories|Accessories&Peripherals|...             ₹154   

  actual_price discount_percentage rating rating_count  \
0       ₹1,099                 64%    4.2       24,269   
1         ₹349                 43%    4.0       43,994   
2       ₹1,899                 90%    3.9        7,928   
3         ₹699                 53%    4.2       94,363   
4         ₹399                 61%    4.2       16,905   

                                       about_product  \
0  High Compatibility : Compatible With iPhone 12...   
1  Compatible with all Type C enabled devices, be...   
2  【 Fast Charger& Data Sync】-With built-in safet...   
3  The boAt Deuce USB 300 2 in 1 cable is compati...   
4  [CHARGE & SYNC FUNCTION]- This cable comes wit...   

                                             user_id  \
0  AG3D6O4STAQKAY2UVGEUV46KN35Q,AHMY5CWJMMK5BJRBB...   
1  AECPFYFQVRUWC3KGNLJIOREFP5LQ,AGYYVPDD7YG7FYNBX...   
2  AGU3BBQ2V2DDAMOAKGFAWDDQ6QHA,AESFLDV2PT363T2AQ...   
3  AEWAZDZZJLQUYVOVGBEUKSLXHQ5A,AG5HTSFRRE6NL3M5S...   
4  AE3Q6KSUK5P75D5HFYHCRAOLODSA,AFUGIFH5ZAFXRDSZH...   

                                           user_name  \
0  Manav,Adarsh gupta,Sundeep,S.Sayeed Ahmed,jasp...   
1  ArdKn,Nirbhay kumar,Sagar Viswanathan,Asp,Plac...   
2  Kunal,Himanshu,viswanath,sai niharka,saqib mal...   
3  Omkar dhale,JD,HEMALATHA,Ajwadh a.,amar singh ...   
4  rahuls6099,Swasat Borah,Ajay Wadke,Pranali,RVK...   

                                           review_id  \
0  R3HXWT0LRP0NMF,R2AJM3LFTLZHFO,R6AQJGUP6P86,R1K...   
1  RGIQEG07R9HS2,R1SMWZQ86XIN8U,R2J3Y1WL29GWDE,RY...   
2  R3J3EQQ9TZI5ZJ,R3E7WBGK7ID0KV,RWU79XKQ6I1QF,R2...   
3  R3EEUZKKK9J36I,R3HJVYCLYOY554,REDECAZ7AMPQC,R1...   
4  R1BP4L2HH9TFUP,R16PVJEXKV6QZS,R2UPDB81N66T4P,R...   

                                        review_title  \
0  Satisfied,Charging is really fast,Value for mo...   
1  A Good Braided Cable for Your Type C Device,Go...   
2  Good speed for earlier versions,Good Product,W...   
3  Good product,Good one,Nice,Really nice product...   
4  As good as original,Decent,Good one for second...   

                                      review_content  \
0  Looks durable Charging is fine tooNo complains...   
1  I ordered this cable to connect my phone to An...   
2  Not quite durable and sturdy,https://m.media-a...   
3  Good product,long wire,Charges good,Nice,I bou...   
4  Bought this instead of original apple, does th...   

                                            img_link  \
0  https://m.media-amazon.com/images/W/WEBP_40237...   
1  https://m.media-amazon.com/images/W/WEBP_40237...   
2  https://m.media-amazon.com/images/W/WEBP_40237...   
3  https://m.media-amazon.com/images/I/41V5FtEWPk...   
4  https://m.media-amazon.com/images/W/WEBP_40237...   

                                        product_link  
0  https://www.amazon.in/Wayona-Braided-WN3LG1-Sy...  
1  https://www.amazon.in/Ambrane-Unbreakable-Char...  
2  https://www.amazon.in/Sounce-iPhone-Charging-C...  
3  https://www.amazon.in/Deuce-300-Resistant-Tang...  
4  https://www.amazon.in/Portronics-Konnect-POR-1...

In [4]:
# para ver si hay algún valor extraño en el dataframe, Analizo las estadísticas descriptivas
amazon.describe()

product_id                                       product_name  \
count         1465                                               1465   
unique        1351                                               1337   
top     B07JW9H4J1  Fire-Boltt Ninja Call Pro Plus 1.83" Smart Wat...   
freq             3                                                  5   

                                                 category discounted_price  \
count                                                1465             1465   
unique                                                211              550   
top     Computers&Accessories|Accessories&Peripherals|...             ₹199   
freq                                                  233               53   

       actual_price discount_percentage rating rating_count  \
count          1465                1465   1465         1463   
unique          449                  92     28         1143   
top            ₹999                 50%    4.1        9,378   
freq            120                  56    244            9   

                                            about_product  \
count                                                1465   
unique                                               1293   
top     [CHARGE & SYNC FUNCTION]- This cable comes wit...   
freq                                                    6   

                                                  user_id  \
count                                                1465   
unique                                               1194   
top     AHIKJUDTVJ4T6DV6IUGFYZ5LXMPA,AE55KTFVNXYFD5FPY...   
freq                                                   10   

                                                user_name  \
count                                                1465   
unique                                               1194   
top     $@|\|TO$|-|,Sethu madhav,Akash Thakur,Burger P...   
freq                                                   10   

                                                review_id  \
count                                                1465   
unique                                               1194   
top     R3F4T5TRYPTMIG,R3DQIEC603E7AY,R1O4Z15FD40PV5,R...   
freq                                                   10   

                                             review_title  \
count                                                1465   
unique                                               1194   
top     Worked on iPhone 7 and didn’t work on XR,Good ...   
freq                                                   10   

                                           review_content  \
count                                                1465   
unique                                               1212   
top     I am not big on camera usage, personally. I wa...   
freq                                                    8   

                                                 img_link  \
count                                                1465   
unique                                               1412   
top     https://m.media-amazon.com/images/I/413sCRKobN...   
freq                                                    3   

                                             product_link  
count                                                1465  
unique                                               1465  
top     https://www.amazon.in/Wayona-Braided-WN3LG1-Sy...  
freq                                                    1

In [5]:
# para mostrar qué tipos de datos hay en cada columna y cuántos valores no nulos hay, utilizo  el comando .info()
amazon.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1465 entries, 0 to 1464
Data columns (total 16 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   product_id           1465 non-null   object
 1   product_name         1465 non-null   object
 2   category             1465 non-null   object
 3   discounted_price     1465 non-null   object
 4   actual_price         1465 non-null   object
 5   discount_percentage  1465 non-null   object
 6   rating               1465 non-null   object
 7   rating_count         1463 non-null   object
 8   about_product        1465 non-null   object
 9   user_id              1465 non-null   object
 10  user_name            1465 non-null   object
 11  review_id            1465 non-null   object
 12  review_title         1465 non-null   object
 13  review_content       1465 non-null   object
 14  img_link             1465 non-null   object
 15  product_link         1465 non-null   object
dtypes: obj

Las columnas en cuestión están actualmente definidas como **objeto**, lo cual es inapropiado para datos numéricos como

**discounted_price**,

**actual_price**,

**discount_percentage**,

**rating** y **rating_count**.

Estas columnas deberían ser del tipo **float64**. Al intentar realizar la conversión de estas columnas a **float64**, se produjo un **error** debido a la presencia de **símbolos** como comas, signos de moneda y porcentajes. Para solucionar este problema, es necesario **eliminar estos símbolos** antes de cambiar el tipo de dato.




**"Feature engineering"**

**Eliminar símbolos de las columnas y convertirlas en el tipo de datos float64.**

- **Conversion de las columnas discounted_price y actual_price**

In [6]:
# para eliminar las comas y ₹ de las columnas discounted_price y actual_price, utilizo  el comando .str.replace()

amazon['discounted_price'] = amazon['discounted_price'].str.replace("₹",'').str.replace(",",'').astype('float64')

# para convertir el tipo de dato en las columnas discounted_price y actual_price a float64, utilizo el comando .astype()

amazon['actual_price'] = amazon['actual_price'].str.replace("₹",'').str.replace(",",'').astype('float64')

Tanto los precios de descuento como los reales están en Rupias.

 **Convierto las Rupias indias a Euros**. El tipo de cambio de Rupias a Euros es de **0,012.**

In [7]:
# para aplicar una función a cada fila, utilizo el comando .apply()
# utilizo lambda para crear una función.

exchange_rate = 0.012
amazon['discounted_price'] = amazon['discounted_price'].apply(lambda x: x * exchange_rate)
amazon['actual_price'] = amazon['actual_price'].apply(lambda x: x * exchange_rate)

# para redondear los precios a 2 decimales, utilizo el comando round()

amazon['discounted_price'] = round(amazon['discounted_price'], 2)
amazon['actual_price'] = round(amazon['actual_price'], 2)

- **Creación de la columna discount_amount.**

Es útil conocer tanto el **valor exacto de los descuentos** como el **porcentaje** correspondiente.

In [8]:
amazon['discount_amount'] = amazon['actual_price'] - amazon['discounted_price']

- **Conversión de la columna discount_percentage.**

In [9]:
# para eliminar el % de la columna discount_percentage, utilizo el comando .str.replace()
# para convertir el tipo de dato en la columna discount_percentage a float64, utilizo el comando .astype()

amazon['discount_percentage'] = amazon['discount_percentage'].str.replace('%','').astype('float64')

# Divido discount_percentage por 100 para que los valores estén en decimales.

amazon['discount_percentage'] = amazon['discount_percentage'] / 100


- **Conversión de la columna rating.**

Aunque la **columna de calificación** parece correcta. Podría haber algún **valor incompatible** con el tipo de datos **float64**, lo que causa un **error** al intentar cambiar el tipo de datos a **float64**.



In [10]:
# para ver los valores únicos en una columna y cuántos de cada valor hay, utilizo el comando .value_counts()

amazon['rating'].value_counts()


rating
4.1    244
4.3    230
4.2    228
4.0    129
3.9    123
4.4    123
3.8     86
4.5     75
4       52
3.7     42
3.6     35
3.5     26
4.6     17
3.3     16
3.4     10
4.7      6
3.1      4
5.0      3
3.0      3
4.8      3
3.2      2
2.8      2
2.3      1
|        1
2        1
3        1
2.6      1
2.9      1
Name: count, dtype: int64

En la parte inferior de la lista, se encuentra un **símbolo "|"** que bloquea la conversión de los datos al tipo **float64**. Para resolver este problema, identificaré la **fila afectada** en la columna de **calificación**, verificaré la **calificación correcta** en la **página web del producto**, y sustituiré el **símbolo "|"** por el valor adecuado.


In [11]:
# para seleccionar la(s) fila(s) donde el valor de rating = |, utilizo el comando .loc[]

amazon.loc[amazon['rating'] == '|']


product_id                                       product_name  \
1279  B08L12N5H1  Eureka Forbes car Vac 100 Watts Powerful Sucti...   

                                               category  discounted_price  \
1279  Home&Kitchen|Kitchen&HomeAppliances|Vacuum,Cle...             25.19   

      actual_price  discount_percentage rating rating_count  \
1279         29.99                 0.16      |          992   

                                          about_product  \
1279  No Installation is provided for this product|1...   

                                                user_id  \
1279  AGTDSNT2FKVYEPDPXAA673AIS44A,AER2XFSWNN4LAUCJ5...   

                                              user_name  \
1279  Divya,Dr Nefario,Deekshith,Preeti,Prasanth R,P...   

                                              review_id  \
1279  R2KKTKM4M9RDVJ,R1O692MZOBTE79,R2WRSEWL56SOS4,R...   

                                           review_title  \
1279  Decent product,doesn't pick up sand,Ok ok,Must...   

                                         review_content  \
1279  Does the job well,doesn't work on sand. though...   

                                               img_link  \
1279  https://m.media-amazon.com/images/W/WEBP_40237...   

                                           product_link  discount_amount  
1279  https://www.amazon.in/Eureka-Forbes-Vacuum-Cle...              4.8

La página web del producto indica que la **calificación promedio del producto es 3.9.**

In [12]:
# para reemplazar | con 3.9 en la columna de calificación, utilizo el comando .str.replace()
# para convertir el tipo de dato en la columna de calificación a float64, utilizo el comando .astype()

amazon['rating'] = amazon['rating'].str.replace('|', '3.9').astype('float64')


- **Conversión de la columna rating_count.**

In [13]:
# para eliminar las comas de la columna rating_count, utilizo el comando .str.replace()
# para convertir el tipo de dato en la columna rating_count a float64, utilizo el comando .astype()

amazon['rating_count'] = amazon['rating_count'].str.replace(',', '').astype('float64')


**"Missing Values".**

**Eliminar valores nulos.**

Para asegurar la **integridad de los datos**, es necesario **eliminar los valores nulos** o **reemplazarlos** con un **valor promedio o mediano**.







In [14]:
# para ver cuántos valores nulos hay en cada columna, utilizo el comando .isna() y .sum()

amazon.isna().sum()


product_id             0
product_name           0
category               0
discounted_price       0
actual_price           0
discount_percentage    0
rating                 0
rating_count           2
about_product          0
user_id                0
user_name              0
review_id              0
review_title           0
review_content         0
img_link               0
product_link           0
discount_amount        0
dtype: int64

En la columna rating_count se encuentran **2 valores nulos**

In [15]:
# para encontrar el número de índice de las filas con 1 o más valores nulos, utilizo el comando .index
# para mostrar filas basadas en el número de índice, utilizoel comando .iloc

amazon.iloc[amazon[(amazon.isna().sum(axis = 1) >= 1)].index]


product_id                                       product_name  \
282  B0B94JPY2N  Amazon Brand - Solimo 65W Fast Charging Braide...   
324  B0BQRJ3C47  REDTECH USB-C to Lightning Cable 3.3FT, [Apple...   

                                              category  discounted_price  \
282  Computers&Accessories|Accessories&Peripherals|...              2.39   
324  Computers&Accessories|Accessories&Peripherals|...              2.99   

     actual_price  discount_percentage  rating  rating_count  \
282         11.99                 0.80     3.0           NaN   
324         11.99                 0.75     5.0           NaN   

                                         about_product  \
282  USB C to C Cable: This cable has type C connec...   
324  💎[The Fastest Charge] - This iPhone USB C cabl...   

                          user_id    user_name      review_id  \
282  AE7CFHY23VAJT2FI4NZKKP6GS2UQ       Pranav   RUB7U91HVZ30   
324  AGJC5O5H5BBXWUV7WRIEIOOR3TVQ  Abdul Gafur  RQXD5SAMMPC6L   

                                     review_title  \
282  The cable works but is not 65W as advertised   
324                               Awesome Product   

                                        review_content  \
282  I have a pd supported car charger and I bought...   
324  Quick delivery.Awesome ProductPacking was good...   

                                              img_link  \
282  https://m.media-amazon.com/images/W/WEBP_40237...   
324  https://m.media-amazon.com/images/I/31-q0xhaTA...   

                                          product_link  discount_amount  
282  https://www.amazon.in/Amazon-Brand-Charging-Su...              9.6  
324  https://www.amazon.in/REDTECH-Lightning-Certif...              9.0

**Reemplazo los valores nulos por el promedio de rating_count.**

In [16]:
# para reemplazar los valores nulos con el valor promedio redondeado, utilizo el comando .fillna()
# para redondear el promedio al número entero más cercano, utilizo el comando round()
# para calcular el oromedio de rating_count, utilizo el comando .mean()

amazon['rating_count'] = amazon['rating_count'].fillna(round(amazon["rating_count"].mean()))


In [17]:
# para ver cuántos valores nulos hay en cada columna, utilizo el comando .isna().sum()

amazon.isna().sum()


product_id             0
product_name           0
category               0
discounted_price       0
actual_price           0
discount_percentage    0
rating                 0
rating_count           0
about_product          0
user_id                0
user_name              0
review_id              0
review_title           0
review_content         0
img_link               0
product_link           0
discount_amount        0
dtype: int64

**"Feature engineering"**

**Eliminar valores duplicados.**


In [18]:
# para eliminar filas duplicadas, utilizo el comando .drop_duplicates()

amazon = amazon.drop_duplicates()


**Eliminar espacio en blanco de product_id.**


In [19]:
# utilizo el comando .str.strip() para eliminar cualquier espacio en blanco al principio y al final de una cadena.

amazon['product_id'].str.strip()


0       B07JW9H4J1
1       B098NS6PVG
2       B096MSW6CT
3       B08HDJ86NZ
4       B08CF3B7N1
           ...    
1460    B08L7J3T31
1461    B01M6453MB
1462    B009P2LIL4
1463    B00J5DYCCA
1464    B01486F4G6
Name: product_id, Length: 1465, dtype: object

**Separar múltiples categorías en varias columnas.**


Es necesario dividir cada categoría en **columnas individuales** y ajustar los nombres para que sean más **claros y legibles**. Las categorías en la columna correspondiente están separadas por el símbolo **'|'** (como en **Ordenadores y accesorios | Accesorios y periféricos**). Para hacer estos datos más manejables en el análisis.


In [20]:
# utilizo el comando .str.split() para dividir los valores de cadena en la columna de categoría usando | como delimitador.
# expand = True insertará cada elemento dividido en una columna separada.

splitcategory = amazon['category'].str.split('|', expand = True)
splitcategory


0                           1  \
0     Computers&Accessories     Accessories&Peripherals   
1     Computers&Accessories     Accessories&Peripherals   
2     Computers&Accessories     Accessories&Peripherals   
3     Computers&Accessories     Accessories&Peripherals   
4     Computers&Accessories     Accessories&Peripherals   
...                     ...                         ...   
1460           Home&Kitchen      Kitchen&HomeAppliances   
1461           Home&Kitchen      Kitchen&HomeAppliances   
1462           Home&Kitchen  Heating,Cooling&AirQuality   
1463           Home&Kitchen  Heating,Cooling&AirQuality   
1464           Home&Kitchen      Kitchen&HomeAppliances   

                               2                         3          4     5  \
0             Cables&Accessories                    Cables  USBCables  None   
1             Cables&Accessories                    Cables  USBCables  None   
2             Cables&Accessories                    Cables  USBCables  None   
3             Cables&Accessories                    Cables  USBCables  None   
4             Cables&Accessories                    Cables  USBCables  None   
...                          ...                       ...        ...   ...   
1460  WaterPurifiers&Accessories  WaterPurifierAccessories       None  None   
1461      SmallKitchenAppliances         Rice&PastaCookers       None  None   
1462                 RoomHeaters            HeatConvectors       None  None   
1463                        Fans               ExhaustFans       None  None   
1464      SmallKitchenAppliances            SandwichMakers       None  None   

         6  
0     None  
1     None  
2     None  
3     None  
4     None  
...    ...  
1460  None  
1461  None  
1462  None  
1463  None  
1464  None  

[1465 rows x 7 columns]

In [21]:
# utilizo el comando .isna().sum() para ver cuántos valores nulos hay en cada columna.

splitcategory.isna().sum()


0       0
1       0
2       8
3     165
4     943
5    1380
6    1452
dtype: int64

En el análisis, solo consideraremos las **columnas 0 y 1** para evitar la presencia de **valores nulos** en las filas sin necesidad de eliminarlas.


In [22]:
# utilizo el comando .rename() para renombrar la columna 0 a "category" y la columna 1 a "subcategory".

splitcategory = splitcategory.rename(columns = {0:'category',
                                                1:'subcategory'})
splitcategory


category                 subcategory  \
0     Computers&Accessories     Accessories&Peripherals   
1     Computers&Accessories     Accessories&Peripherals   
2     Computers&Accessories     Accessories&Peripherals   
3     Computers&Accessories     Accessories&Peripherals   
4     Computers&Accessories     Accessories&Peripherals   
...                     ...                         ...   
1460           Home&Kitchen      Kitchen&HomeAppliances   
1461           Home&Kitchen      Kitchen&HomeAppliances   
1462           Home&Kitchen  Heating,Cooling&AirQuality   
1463           Home&Kitchen  Heating,Cooling&AirQuality   
1464           Home&Kitchen      Kitchen&HomeAppliances   

                               2                         3          4     5  \
0             Cables&Accessories                    Cables  USBCables  None   
1             Cables&Accessories                    Cables  USBCables  None   
2             Cables&Accessories                    Cables  USBCables  None   
3             Cables&Accessories                    Cables  USBCables  None   
4             Cables&Accessories                    Cables  USBCables  None   
...                          ...                       ...        ...   ...   
1460  WaterPurifiers&Accessories  WaterPurifierAccessories       None  None   
1461      SmallKitchenAppliances         Rice&PastaCookers       None  None   
1462                 RoomHeaters            HeatConvectors       None  None   
1463                        Fans               ExhaustFans       None  None   
1464      SmallKitchenAppliances            SandwichMakers       None  None   

         6  
0     None  
1     None  
2     None  
3     None  
4     None  
...    ...  
1460  None  
1461  None  
1462  None  
1463  None  
1464  None  

[1465 rows x 7 columns]

In [23]:
# utilizo el comando .unique() para obtener una lista de valores únicos en la columna de categoría.

splitcategory['category'].unique()


array(['Computers&Accessories', 'Electronics', 'MusicalInstruments',
       'OfficeProducts', 'Home&Kitchen', 'HomeImprovement', 'Toys&Games',
       'Car&Motorbike', 'Health&PersonalCare'], dtype=object)

In [24]:
# Formateo los valores de la categoría para que sean más fáciles de leer.
# Uso .str.replace() para agregar espacios a los valores de la categoría.

splitcategory['category'] = splitcategory['category'].str.replace('&',
                                                                  ' & ')

splitcategory['category'] = splitcategory['category'].str.replace('MusicalInstruments',
                                                                  'Musical Instruments')

splitcategory['category'] = splitcategory['category'].str.replace('OfficeProducts',
                                                                  'Office Products')

splitcategory['category'] = splitcategory['category'].str.replace('HomeImprovement',
                                                                  'Home Improvement')


In [25]:
# utilizoel comando .unique() para obtener una lista de valores únicos en la columna de subcategoría.

splitcategory['subcategory'].unique()


array(['Accessories&Peripherals', 'NetworkingDevices',
       'HomeTheater,TV&Video', 'HomeAudio', 'WearableTechnology',
       'Mobiles&Accessories', 'Accessories',
       'Headphones,Earbuds&Accessories', 'ExternalDevices&DataStorage',
       'Microphones', 'GeneralPurposeBatteries&BatteryChargers',
       'OfficePaperProducts', 'CraftMaterials', 'Cameras&Photography',
       'OfficeElectronics', 'Printers,Inks&Accessories', 'Monitors',
       'Components', 'Electrical', 'Arts&Crafts', 'PowerAccessories',
       'Tablets', 'Laptops', 'Kitchen&HomeAppliances',
       'Heating,Cooling&AirQuality', 'Kitchen&Dining',
       'HomeStorage&Organization', 'CarAccessories',
       'HomeMedicalSupplies&Equipment'], dtype=object)

In [26]:
# utilizo .str.replace() para agregar espacios a los valores de la subcategoría.

splitcategory['subcategory'] = splitcategory['subcategory'].str.replace('&',
                                                                        ' & ')

splitcategory['subcategory'] = splitcategory['subcategory'].str.replace(',',
                                                                        ', ')

splitcategory['subcategory'] = splitcategory['subcategory'].str.replace('NetworkingDevices',
                                                                        'Networking Devices')

splitcategory['subcategory'] = splitcategory['subcategory'].str.replace('HomeTheater',
                                                                        'Home Theater')

splitcategory['subcategory'] = splitcategory['subcategory'].str.replace('HomeAudio',
                                                                        'Home Audio')

splitcategory['subcategory'] = splitcategory['subcategory'].str.replace('WearableTechnology',
                                                                        'Wearable Technology')

splitcategory['subcategory'] = splitcategory['subcategory'].str.replace('ExternalDevices',
                                                                        'External Devices')

splitcategory['subcategory'] = splitcategory['subcategory'].str.replace('DataStorage',
                                                                        'Data Storage')

splitcategory['subcategory'] = splitcategory['subcategory'].str.replace('GeneralPurposeBatteries',
                                                                        'General Purpose Batteries')

splitcategory['subcategory'] = splitcategory['subcategory'].str.replace('BatteryChargers',
                                                                        'Battery Chargers')

splitcategory['subcategory'] = splitcategory['subcategory'].str.replace('OfficePaperProducts',
                                                                        'Office Paper Products')

splitcategory['subcategory'] = splitcategory['subcategory'].str.replace('CraftMaterials',
                                                                        'Craft Materials')

splitcategory['subcategory'] = splitcategory['subcategory'].str.replace('OfficeElectronics',
                                                                        'Office Electronics')

splitcategory['subcategory'] = splitcategory['subcategory'].str.replace('PowerAccessories',
                                                                        'Power Accessories')

splitcategory['subcategory'] = splitcategory['subcategory'].str.replace('HomeAppliances',
                                                                        'Home Appliances')

splitcategory['subcategory'] = splitcategory['subcategory'].str.replace('AirQuality',
                                                                        'Air Quality')

splitcategory['subcategory'] = splitcategory['subcategory'].str.replace('HomeStorage',
                                                                        'Home Storage')

splitcategory['subcategory'] = splitcategory['subcategory'].str.replace('CarAccessories',
                                                                        'Car Accessories')

splitcategory['subcategory'] = splitcategory['subcategory'].str.replace('HomeMedicalSupplies',
                                                                        'Home Medical Supplies')


In [27]:
# utilizo el comando .drop() para eliminar la columna de categoría del dataframe de amazon.

amazon = amazon.drop(columns = 'category')

# Agrego la nueva columna de categoría y la columna de subcategoría al dataframe de amazon.
amazon['category'] = splitcategory['category']
amazon['subcategory'] = splitcategory['subcategory']


**Creo una columna que clasifique los productos en función de la puntuación de calificación.**

Cada rango numérico corresponderá a una **clasificación específica** (medidas de: **0 - 0.9 estrellas = Muy pobre** a **5 estrellas = Excelente**). La **columna de clasificación** se basará en las **calificaciones de los productos**.



In [28]:
# Creo un bucle for y una declaración if else para categorizar la escala de calificación.
# utilizo el comando .append() para agregar categorías de clasificación a una lista de clasificación.

ranking = []

for score in amazon['rating']:
    if score <= 0.9: ranking.append('Muy Malo')
    elif score <= 1.9: ranking.append('Malo')
    elif score <= 2.9: ranking.append('Promedio')
    elif score <= 3.9: ranking.append('Bueno')
    elif score <= 4.9: ranking.append('Muy Bueno')
    elif score == 5.0: ranking.append('Excelente')

# Agrego la lista de clasificación como una columna de clasificación en el dataframe de amazon.
# utilizo el comando .astype() para convertir los valores en la columna de clasificación en un tipo de dato de categoría.

amazon['ranking'] = ranking

amazon['ranking'] = amazon['ranking'].astype('category')


**Creo un dataframe con columnas que se utilizarán en el análisis.**

Se ha creado un **nuevo dataframe** que solo conserva las **columnas esenciales** para el análisis, tras la limpieza de los datos. Algunas columnas, como **img_link** y **product_link**, no se incluirán en el análisis.



In [29]:
amazon_clean = amazon[['product_id',
                       'product_name',
                       'actual_price',
                       'discounted_price',
                       'discount_amount',
                       'discount_percentage',
                       'category',
                       'subcategory',
                       'rating',
                       'rating_count',
                       'ranking']]
amazon_clean

product_id                                       product_name  \
0     B07JW9H4J1  Wayona Nylon Braided USB to Lightning Fast Cha...   
1     B098NS6PVG  Ambrane Unbreakable 60W / 3A Fast Charging 1.5...   
2     B096MSW6CT  Sounce Fast Phone Charging Cable & Data Sync U...   
3     B08HDJ86NZ  boAt Deuce USB 300 2 in 1 Type-C & Micro USB S...   
4     B08CF3B7N1  Portronics Konnect L 1.2M Fast Charging 3A 8 P...   
...          ...                                                ...   
1460  B08L7J3T31  Noir Aqua - 5pcs PP Spun Filter + 1 Spanner | ...   
1461  B01M6453MB  Prestige Delight PRWO Electric Rice Cooker (1 ...   
1462  B009P2LIL4  Bajaj Majesty RX10 2000 Watts Heat Convector R...   
1463  B00J5DYCCA  Havells Ventil Air DSP 230mm Exhaust Fan (Pist...   
1464  B01486F4G6  Borosil Jumbo 1000-Watt Grill Sandwich Maker (...   

      actual_price  discounted_price  discount_amount  discount_percentage  \
0            13.19              4.79             8.40                 0.64   
1             4.19              2.39             1.80                 0.43   
2            22.79              2.39            20.40                 0.90   
3             8.39              3.95             4.44                 0.53   
4             4.79              1.85             2.94                 0.61   
...            ...               ...              ...                  ...   
1460         11.03              4.55             6.48                 0.59   
1461         36.54             27.36             9.18                 0.25   
1462         36.96             26.63            10.33                 0.28   
1463         22.68             16.79             5.89                 0.26   
1464         44.28             34.36             9.92                 0.22   

                     category                     subcategory  rating  \
0     Computers & Accessories       Accessories & Peripherals     4.2   
1     Computers & Accessories       Accessories & Peripherals     4.0   
2     Computers & Accessories       Accessories & Peripherals     3.9   
3     Computers & Accessories       Accessories & Peripherals     4.2   
4     Computers & Accessories       Accessories & Peripherals     4.2   
...                       ...                             ...     ...   
1460           Home & Kitchen       Kitchen & Home Appliances     4.0   
1461           Home & Kitchen       Kitchen & Home Appliances     4.1   
1462           Home & Kitchen  Heating, Cooling & Air Quality     3.6   
1463           Home & Kitchen  Heating, Cooling & Air Quality     4.0   
1464           Home & Kitchen       Kitchen & Home Appliances     4.3   

      rating_count    ranking  
0          24269.0  Muy Bueno  
1          43994.0  Muy Bueno  
2           7928.0      Bueno  
3          94363.0  Muy Bueno  
4          16905.0  Muy Bueno  
...            ...        ...  
1460        1090.0  Muy Bueno  
1461        4118.0  Muy Bueno  
1462         468.0      Bueno  
1463        8031.0  Muy Bueno  
1464        6987.0  Muy Bueno  

[1465 rows x 11 columns]

In [30]:
# Guardar el DataFrame limpio en un archivo CSV
amazon_clean.to_csv('/content/amazon_clean.csv', index=False)


Para garantizar la calidad y consistencia de los datos, este enfoque asegura que el **dataframe amazon_clean** esté preparado de manera óptima para el análisis posterior. Al mismo tiempo, se han eliminado las columnas que no se considerarán en el análisis, optimizando así la estructura del dataframe y mejorando la eficiencia del proceso analítico. Se han agregado nuevas columnas que proporcionan información adicional relevante para el análisis. Las columnas que contenían más de un valor se han dividido de manera que cada valor ahora tenga su propia columna, facilitando un análisis más preciso y detallado.

Los datos del **dataframe amazon_clean** se han formateado meticulosamente para asegurar que cada columna tenga los tipos de datos apropiados. El **dataframe limpio** tiene **1465 filas**, coincidiendo con la cantidad de filas del **dataframe original y no limpiado**.




**Creo un dataframe con reviewer data.**

Para permitir un análisis más detallado de los **revisores**, se necesita una estructura adicional, ya que el **dataframe amazon_clean** se enfoca en los productos vendidos por Amazon. Es necesario un **dataframe adicional** que incluya los datos de los revisores para analizar a los **clientes que revisaron los productos**. Actualmente, las columnas **user_name** y **user_id** contienen múltiples valores separados por comas; estos valores deben dividirse en **filas individuales** en lugar de columnas.



In [31]:
# utilizo el comando .str.split() para dividir los valores de cadena en las columnas user_id y user_name usando ',' como delimitador.
# expand = False insertará cada elemento dividido en una lista.

split_user_id = amazon['user_id'].str.split(',', expand = False)
split_user_name = amazon['user_name'].str.split(',', expand = False)


In [32]:
# utilizo el comando .explode() para dividir cada elemento de una lista en una fila.
# Nota: Aunque cada elemento de la lista está en una fila diferente, los elementos comparten el mismo número de índice.
# Ejemplo: Una lista con 5 valores se dividirá en 5 filas, pero cada una de esas 5 filas tendrá el mismo número de índice.

id_rows = split_user_id.explode()
name_rows = split_user_name.explode()


In [33]:
# utilizo el comando DataFrame() para crear un dataframe usando the exploded lists.

df_id_rows = pd.DataFrame(id_rows)
df_name_rows = pd.DataFrame(name_rows)


In [34]:
# Agrego las columnas product_name, category y subcategory del dataframe amazon_clean al dataframe df_name_rows.

df_name_rows['product_name'] = amazon_clean['product_name']
df_name_rows['category'] = amazon_clean['category']
df_name_rows['subcategory'] = amazon_clean['subcategory']


In [35]:
# utilizo el comando .reset_index() para restablecer el índice de modo que cada fila tenga su propio número de índice.

df_id_rows = df_id_rows.reset_index(drop = True)
df_name_rows = df_name_rows.reset_index(drop = True)


In [36]:
# utilizo el comando .merge() para fusionar 2 dataframes juntos.

reviewers = pd.merge(df_id_rows, df_name_rows, left_index = True, right_index = True)
reviewers


user_id          user_name  \
0      AG3D6O4STAQKAY2UVGEUV46KN35Q              Manav   
1      AHMY5CWJMMK5BJRBBSNLYT3ONILA       Adarsh gupta   
2      AHCTC6ULH4XB6YHDY6PCH2R772LQ            Sundeep   
3      AGYHHIERNXKA6P5T7CZLXKVPT7IQ     S.Sayeed Ahmed   
4      AG4OGOFWXJZTQ2HKYIOCOY3KXF2Q     jaspreet singh   
...                             ...                ...   
11498  AHXCDNSXAESERITAFELQABFVNLCA  BATTU SURESHKUMAR   
11499  AGRZD6CHLCUNOLMMIMIHUCG7PIFA      Shubham Dubey   
11500  AFQZVGSOSOJHKFQQMCEI4725QEKQ        E.GURUBARAN   
11501  AEALVGXXIP46OZVXKRUXSDWZJMEA          Mayank S.   
11502  AGEFL3AY7YXEFZA4ZJU3LP7K7OJQ         eusuf khan   

                                            product_name  \
0      Wayona Nylon Braided USB to Lightning Fast Cha...   
1      Wayona Nylon Braided USB to Lightning Fast Cha...   
2      Wayona Nylon Braided USB to Lightning Fast Cha...   
3      Wayona Nylon Braided USB to Lightning Fast Cha...   
4      Wayona Nylon Braided USB to Lightning Fast Cha...   
...                                                  ...   
11498  Bajaj Majesty RX10 2000 Watts Heat Convector R...   
11499  Havells Ventil Air DSP 230mm Exhaust Fan (Pist...   
11500  Havells Ventil Air DSP 230mm Exhaust Fan (Pist...   
11501  Havells Ventil Air DSP 230mm Exhaust Fan (Pist...   
11502  Havells Ventil Air DSP 230mm Exhaust Fan (Pist...   

                      category                     subcategory  
0      Computers & Accessories       Accessories & Peripherals  
1      Computers & Accessories       Accessories & Peripherals  
2      Computers & Accessories       Accessories & Peripherals  
3      Computers & Accessories       Accessories & Peripherals  
4      Computers & Accessories       Accessories & Peripherals  
...                        ...                             ...  
11498           Home & Kitchen  Heating, Cooling & Air Quality  
11499           Home & Kitchen  Heating, Cooling & Air Quality  
11500           Home & Kitchen  Heating, Cooling & Air Quality  
11501           Home & Kitchen  Heating, Cooling & Air Quality  
11502           Home & Kitchen  Heating, Cooling & Air Quality  

[11503 rows x 5 columns]

In [37]:
# utilizo el comando .isna().sum() para ver cuántos valores nulos hay en el dataframe reviewers.

reviewers.isna().sum()


user_id         0
user_name       0
product_name    0
category        0
subcategory     0
dtype: int64

In [38]:
# Guardar el DataFrame reviewers en un archivo CSV
reviewers.to_csv('reviewers_data.csv', index=False)



**creación de `amazon_dataset_final.csv` para el Modelo Regresion Lineal**

Este proceso tiene como objetivo generar un archivo CSV, llamado `amazon_dataset_final.csv`, que contiene solo las columnas necesarias para entrenar el modelo Reresion Lineal

**Descripción de las columnas**:
   - `actual_price`: Precio original del producto.
   - `discounted_price`: Precio del producto con descuento aplicado.
   - `discount_percentage`: Porcentaje de descuento.
   - `rating`: Calificación promedio del producto.
   - `rating_count`: Cantidad de valoraciones que ha recibido el producto.


In [39]:
import pandas as pd

# Cargar el archivo de origen
amazon_clean = pd.read_csv('/content/amazon_clean.csv')

# Seleccionar las columnas requeridas en el orden especificado
columns_to_select = ['actual_price', 'discounted_price', 'discount_percentage', 'rating', 'rating_count']
filtered_data = amazon_clean[columns_to_select]

# Guardar el resultado en un archivo CSV
filtered_data.to_csv('/content/amazon_dataset_final.csv', index=False)
